editor/plugins/sprite_2d_editor_plugin.cpp

```diff
@@ -499,10 +492,94 @@ void Sprite2DEditor::_debug_uv_draw() {
+	h_scroll->set_value(offset.x);
+	v_scroll->set_value(offset.y);
+	updating_zoom = false;
+	_update_zoom_and_pan(false);
+}
+
+void Sprite2DEditor::_pan_callback(Vector2 p_scroll_vec, Ref<InputEvent> p_event) {
+	updating_zoom = true;
+	h_scroll->set_value(h_scroll->get_value() - p_scroll_vec.x / draw_zoom);
+	v_scroll->set_value(v_scroll->get_value() - p_scroll_vec.y / draw_zoom);
+	updating_zoom = false;
+	_update_zoom_and_pan(false);
+}
+
+void Sprite2DEditor::_zoom_callback(float p_zoom_factor, Vector2 p_origin, Ref<InputEvent> p_event) {
+	updating_zoom = true;
+	const real_t prev_zoom = draw_zoom;
+	zoom_widget->set_zoom(draw_zoom * p_zoom_factor);
+	draw_offset += p_origin / prev_zoom - p_origin / zoom_widget->get_zoom();
+	h_scroll->set_value(draw_offset.x);
+	v_scroll->set_value(draw_offset.y);
+	updating_zoom = false;
+	_update_zoom_and_pan(false);
+}
+
+void Sprite2DEditor::_update_zoom_and_pan(bool zoom_at_center) {
+	if (updating_zoom) {
+		return;
+	}
+	updating_zoom = true;
+	real_t previous_zoom = draw_zoom;
+	draw_zoom = zoom_widget->get_zoom();
+	draw_offset = Vector2(h_scroll->get_value(), v_scroll->get_value());
+	if (zoom_at_center) {
+		Vector2 center = debug_uv->get_size() / 2;
+		draw_offset += center / previous_zoom - center / draw_zoom;
+	}
+
+	Ref<Texture2D> tex = node->get_texture();
+	ERR_FAIL_COND(!tex.is_valid());
+
+	Point2 min_corner = Point2();
+	Point2 max_corner = tex->get_size();
+	Size2 page_size = debug_uv->get_size() / draw_zoom;
+	Vector2 margin = Vector2(50, 50) * EDSCALE / draw_zoom;
+	min_corner -= page_size - margin;
+	max_corner += page_size - margin;
+
+	h_scroll->set_min(min_corner.x);
+	h_scroll->set_max(max_corner.x);
+	h_scroll->set_page(page_size.x);
+	h_scroll->set_value(draw_offset.x);
+
+	v_scroll->set_min(min_corner.y);
+	v_scroll->set_max(max_corner.y);
+	v_scroll->set_page(page_size.y);
+	v_scroll->set_value(draw_offset.y);
+
+	updating_zoom = false;
+	debug_uv->queue_redraw();
+}
+
 void Sprite2DEditor::_notification(int p_what) {
 	switch (p_what) {
+		case NOTIFICATION_READY: {
+			v_scroll->set_anchors_and_offsets_preset(Control::PRESET_RIGHT_WIDE);
+			h_scroll->set_anchors_and_offsets_preset(Control::PRESET_BOTTOM_WIDE);
+			// Avoid scrollbar overlapping.
+			Size2 hmin = h_scroll->get_combined_minimum_size();
+			Size2 vmin = v_scroll->get_combined_minimum_size();
+			h_scroll->set_anchor_and_offset(SIDE_RIGHT, ANCHOR_END, -vmin.width);
+			v_scroll->set_anchor_and_offset(SIDE_BOTTOM, ANCHOR_END, -hmin.height);
+		} break;
 		case NOTIFICATION_ENTER_TREE:
-		case NOTIFICATION_THEME_CHANGED: {
+		case NOTIFICATION_THEME_CHANGED:
+		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
+			panner->setup((ViewPanner::ControlScheme)EDITOR_GET("editors/panning/sub_editors_panning_scheme").operator int(), ED_GET_SHORTCUT("canvas_item_editor/pan_view"), bool(EDITOR_GET("editors/panning/simple_panning")));
```

Comment:
This shouldn't be happening with `NOTIFICATION_THEME_CHANGED` and the code below shouldn't be happening with `NOTIFICATION_EDITOR_SETTINGS_CHANGED`. You could reorder the cases and use fallthrough.